### PR TITLE
Layout: Make Logged-out Masterbar links Jetpack aware

### PR DESCRIPTION
**Pull request by sirreal**
This PR updates the logged out masterbar links to behave in an appropriate way from the Jetpack connection flow:

- Signup is removed from signup paths (`/jetpack/connect/authorize` and `/jetpack/new`).
- Signup in the `jetpack-connect` section now points to `/jetpack/new`.
- Login links from `jetpack-connect` section use the `log-in/jetpack` login page.
- Preserves email when navigating to log-in from the Jetpack connect signup form.
- Fixes a bug with render race conditions by using redux state over `window.location`.

See p1HpG7-4P7-p2

Extracted from #22218

## Testing

Changes are limited to the Logged Out Masterbar. Ensure you're logged out of WordPress.com. Emails with and without associated WordPress.com accounts should behave in a coherent way. This changes are all here:

this

Clicking "Log In" should always land you at `log-in/jetpack`.
"Sign up" may be hidden depending on where you are (authorize or the new site screen).
"Sign

This should have _no impact at all_ outside of the `jetpack-connect` section, except for improving the "Sign Up" link from `log-in/jetpack`.

1. Connect a new site
1. You can swap out `calypso.live` and add the branch param `branch=update/masterbar-jetpack-login` once you arrive in Calypso for testing.
1. Use the masterbar to navigate between "Log In" (`log-in/jetpack`) and "Sign up" (`/jetpack/connect/authorize`)
1. All good?
1. Try visiting other URLs and check the behavior:
   - https://calypso.live/log-in/jetpack?branch=update/masterbar-jetpack-login
   - https://calypso.live/jetpack/new?branch=update/masterbar-jetpack-login
   - https://calypso.live/jetpack/connect/store?branch=update/masterbar-jetpack-login
   - https://calypso.live/jetpack/connect?branch=update/masterbar-jetpack-login
1. Other logged out URLs unaffected
   - https://calypso.live/log-in?branch=update/masterbar-jetpack-login
   - https://calypso.live/start?branch=update/masterbar-jetpack-login
   - …

### CODE REVIEW

**Comment by sirreal** [comment]
Bringing https://github.com/Automattic/wp-calypso/pull/22218#issuecomment-365672503 over:

> > If you toggle back and forth between login/Jetpack Connect signup via the masterbar, sometimes the redirection is wrong and the masterbar signup points to /start
> 
> It always seems to happen on the toggle _after_ this warning has appeared in the console:
> 
> screen shot 2018-02-14 at 16 54 54

 - @seear 

My intuition says this is from `layout/community-translator/launcher` component. I believe it's unrelated because I can get this warning while maintaining a working link.

**Comment by sirreal** [comment]
Consistently reproducible:

1. Connect from WP Admin, logged out of WordPress.com with a matching WordPress.com account email
1. Land on `log-in/jetpack`
1. Click masterbar "Sign Up"
1. Click masterbar "Log In"
1. Click masterbar "Sign Up"
1. Click masterbar "Log In"
1. "Sign Up" now points to `/start` 

This is reproducible fast or slow, it consistently gives the result.

**Comment by sirreal** [comment]
The issue seems to be a race condition related to `window.location` in render:

1. Connect from WP Admin, logged out of WordPress.com with a matching WordPress.com account email
1. Land on `log-in/jetpack`
1. Click masterbar "Sign Up"
1. Click masterbar "Log In"
1. Click masterbar "Sign Up"
1. Find the masterbar in React devtool and select it
1. Enter `$r.forceUpdate()` in the console to force re-render of masterbar component.
1. Click masterbar "Log In"
1. "Sign Up" now points to `/start`

**Comment by sirreal** [comment]
Bug is fixed by using redux state in the masterbar over window.

File: client/layout/test/index.js [hunk 1]
```diff
@@ -15,15 +16,17 @@ describe( 'index', () => {
 		test( "doesn't throw an exception", () => {
 			expect( () => {
 				renderToString(
-					<LayoutLoggedOut
+					<Provider
```
Comment (sirreal) [inline client/layout/test/index.js]: The prevents connected logged out masterbar from failing tests.

**Comment by rickybanister** [comment]
Clicking 'sign up' from /log-in/jetpack took me to here:

image

Which makes sense, but I didn't expect that. I suppose that's because I navigated directly to the /log-in/jetpack route without a site parameter? 

When I navigated directly to /log-in and then clicked 'sign up' I saw the wpcom nux as I expected. 

I think this is ready to go.

**Comment by sirreal** [comment]
> Which makes sense, but I didn't expect that.

@rickybanister Correct, that's a new change. I chose that for a signup page from within the "jetpack-connect" section.

If we arrive at the `/log-in/jetpack` page with the possibility of continuing a site connect, the signup link should take you to the signup form in the connection flow.

File: client/layout/masterbar/logged-out.jsx [hunk 2]
```diff
@@ -5,70 +5,150 @@
...
+
+		/**
+		 * Hide signup from from New Site screen. This allows starting with a new Jetpack or
+		 * WordPress.com site.
+		 */
+		if ( startsWith( currentRoute, '/jetpack/new' ) ) {
+			return null;
+		}
+
+		let signupUrl = config( 'signup_url' );
```
Comment (seear) [inline client/layout/masterbar/logged-out.jsx]: This makes testing hard because it links to wordpress.com :(

**Comment by seear** [comment]
Working well in production, great work as usual @sirreal.

**Pull request merged**